### DISCUSSION TITLE
Database Provider Packages and Git

### DISCUSSION THREAD

**Discussion by sbwalker** [oqtane/oqtane.framework#3499] [General]
Can someone let me know if after the changes in #3494 and you are using a local clone of the Oqtane.Framework source code... after you run the framework locally, do the packages in the Oqtane.Server/wwwroot/Packages/ folder appear in Visual Studio in the Git Changes tab as being deleted (ie. identified in the Changes area with a D after them). They should not appear in this areas as the .gitignore should ignore *.nuspec packages but I want to make sure that this is working correctly for others.

**Comment by leigh-pointer** [comment]
@sbwalker yes they appear in Visual Studio in the Git Changes tab as being deleted.  Was happening to me today.

**Reply by sbwalker** [reply]
Thank you for letting me know. This is not optimal but I am not sure how to fix it - its essentially a conflict between needing to keep some assets in source control but not wanting them to be tracked. I know that you can use local git commands to instruct the git index to not track them - but those are command line instructions that you need execute on each dev environment. The way it was being managed previously was to not delete the packages from wwwroot/Package - which is not a good solution for the reasons I outlined below.

**Comment by sbwalker** [comment]
To provide some more context.... #3494 changes the approach for database provider packages:

- they are created in wwwroot/Packages so that dot net publish includes them in the release output
- although they are included in release packages (via dot net publish) they still need to be included in source control for the scenario where someone has cloned the framework locally for development - as they will need these packages to be installed to the /bin in order for the framework to run successfully
- they are renamed from *.nuspec.bak to *.nuspec (as the .bak extension was originally used to prevent them from being installed - but is no longer needed)
- they are automatically moved to /Packages on startup as this is a secure location (previously the *.nupkg.bak files were copied but the original files were left in the wwwroot/Packages folder to avoid challenges with source control - however it is not a good practice to leave these assets in a publicly accessible folder after installation... and this was also causing the default database provider to be re-installed everytime the framework started up)
- all database provider *.nuspec packages are installed automatically during startup - very early in the pipeline before assemblies are loaded (this is a change from previous where only the default database provider identified in appsettings would be installed - and it was being installed too late in the process causing assembly conflicts)

I am not 100% happy with these changes, as it means that all database providers are installed even though they are not being used. And the size of the MySQL and PostgreSQL providers are rather large (~1MB) which will be loaded into the app domain on startup. In general the goal in Oqtane is to only load resources which are absolutely required to ensure the fastest possible startup performance. I plan to optimize this further.
